design/crypto_primitives.rs

```diff
@@ -0,0 +1,205 @@
+//! `PublicKeys` and `SecretKeys` into corresponding types: `PublicEncryptKey, `SecretEncryptKey`,
+//! `PublicSignKey` and `SecretSignKey`.
+
+// whatever is defined in the underlying crypto library
+const PUBLIC_ENCRYPT_KEY_BYTES: usize = rust_sodium::crypto::box_::PUBLICKEYBYTES;
+const SECRET_ENCRYPT_KEY_BYTES: usize = rust_sodium::crypto::sign::SECRETKEYBYTES;
+const PUBLIC_SIGN_KEY_BYTES: usize = rust_sodium::crypto::sign::PUBLICKEYBYTES;
+const SECRET_SIGN_KEY_BYTES: usize = rust_sodium::crypto::sign::SECRETKEYBYTES;
+const SHARED_SECRET_KEY_BYTES: usize = XXX;
+const SIGNATURE_BYTES: usize = XXX;
+const SYMMETRIC_KEY_BYTES: usize = XXX;
+
+quick_error! {
+    /// This error is returned if encryption or decryption fail.
+    /// The encryption failure is rare and mostly connected to serialisation failures.
+    /// Decryption can fail because of invalid keys, invalid data, or deserialisation failures.
+    #[derive(Debug)]
+    pub enum Error {
+        /// Occurs when serialisation or deserialisation fails.
+        Serialisation(e: SerialisationError) {
+            description("error serialising or deserialising message")
+            display("error serialising or deserialising message: {}", e)
+            cause(e)
+            from()
+        }
+        /// Occurs when we can't decrypt a message or verify the signature.
+        DecryptVerify(_e: ()) {
+            description("error decrypting/verifying message")
+            from()
+        }
+        /// Occurs in case of an error during initialisation.
+        InitError(e: i32) {
+            description("error while initialising random generator")
+            from()
+        }
+        /// Occurs when we fail to derive encryption key from password.
+        DeriveKey {
+            description("error deriving encryption key from password")
+        }
+    }
+}
+
+/// Parameters to control key derivation which intentionally might be faster or slower depending
+/// on the context.
+pub struct KeyDeriveParams {
+    // some internal parameters
+}
+
+impl KeyDeriveParams {
+    /// Safe base line for CPU and memory parameters for interactive password hashing.
+    pub fn interactive() -> Self
+
+    /// Chooses CPU and memory parameters for highly sensitive data.
+    pub fn sensitive() -> Self
+
+    /// Use custom CPU and memory parameters for key derivation.
+    // Not sure if we want this variantion, cause different implementations have different
+    // parameters. E.g. scrypt uses N, r, p parameters:
+    // https://github.com/DaGenix/rust-crypto/blob/master/src/scrypt.rs#L163
+    pub fn new(cpu_limit: usize, mem_limit: usize) -> Self
+}
+
+impl Default for KeyDeriveParams {
+    fn default() -> Self {
+        KeyDeriveParams::interactive()
+    }
+}
+
+/// Derives encryption key from plaintext password.
+pub fn key_from_password(password: &[u8], salt: &[u8], params: &KeyDeriveParams) -> Result<SymmetricKey, Error>;
```

Comment:
If we don't pass `params` here and use suitable defaults, can this function fail?

Reply:
Mmm, I believe so. Say we still use `rust_sodium` whose docs say:
```
The function returns Ok(key) on success and Err(()) if the computation didn't complete, usually because the operating system refused to allocate the amount of requested memory.
```
https://docs.rs/rust_sodium/0.9.0/rust_sodium/crypto/pwhash/scryptsalsa208sha256/fn.derive_key.html

Sounds like with secure/slow key derivation algorithms this might be a common trait when one has low memory resources.